executor/new_distsql.go

```diff
@@ -286,62 +289,193 @@ type IndexLookUpExecutor struct {
 	ctx       context.Context
 	schema    *expression.Schema
 	// This is the column that represent the handle, we can use handleCol.Index to know its position.
-	handleCol *expression.Column
-
-	// result returns one or more distsql.PartialResult.
-	result distsql.SelectResult
-
-	taskChan chan *lookupTableTask
-	tasksErr error
-	taskCurr *lookupTableTask
-
+	handleCol    *expression.Column
 	tableRequest *tipb.DAGRequest
 	// columns are only required by union scan.
 	columns  []*model.ColumnInfo
 	priority int
+	// All fields above is immutable.
+
+	indexHandler
+	tableHandler
 	finished chan struct{}
 }
 
+// indexHandler is used by IndexLookUpExecutor to maintain index lookup background goroutines.
+type indexHandler struct {
+	buildTableTasks func(handles []int64) []*lookupTableTask
```

Comment:
I think variable function makes the code more complex.
How about just hold a reference to `IndexLookUpExecutor` in `indexHandler`?
So we can just implement `buildTableTasks` directly in `indexHandler`.

Reply:
I've consider that at the first time but abandon the idea.
`IndexLookUpExecutor` include `indexHandler` while `indexHandler` hold a reference to `IndexLookUpExecutor`, it make the code even more clumsy.
As long as the the data fields `buildTableTasks` visited is all immutable, I think using function pointer is affordable.

Reply:
This is a very common pattern, a member holds a reference to its parent.
Except that when the member is embedded.
I think it's better if you make `indexHandler` and `tableHandler` an instance instead of embedded member.

Reply:
Another solution is to add prefix to the functions, like `indexOpen`, `tableOpen`, `indexFetchHandles`.
So we don't need to define `indexHandler` and `tableHandler`.

Use embedded member to separate the logic is not straightforward.

Reply:
In the object-oriented style, we don't care the inner state of the object, just call the methods provide by it. That's a good abstraction for here, I don't want `IndexLookUpExecutor` to cope with a huge couple of inner state, so I split them into smaller object.

Reply:
I don't remember I saw any Go code like this, assign the parent's method to embedded member's function field.
It's like use a very rare way to solve a common problem.